File: backend/core/domain/fen/fen_generator.py

# # ------------------------
# WHITE_PAWN = 0b1001  # 9
# WHITE_KNIGHT = 0b1010  # 10
# WHITE_BISHOP = 0b1011  # 11
# WHITE_ROOK = 0b1100  # 12
# WHITE_QUEEN = 0b1101  # 13
# WHITE_KING = 0b1110  # 14
# # ------------------------
# BLACK_PAWN = 0b0001  # 1
# BLACK_KNIGHT = 0b0010  # 2
# BLACK_BISHOP = 0b0011  # 3
# BLACK_ROOK = 0b0100  # 4
# BLACK_QUEEN = 0b0101  # 5
# BLACK_KING = 0b0110  # 6
# # ------------------------


from core.domain.engine.enums import PieceColor, CastleEnum
from core.domain.engine.Position import Position
from core.domain.engine.square_helping_functions import get_num_by_square_name, get_bitboard_from_num


def set_piece_placement(position: Position, pieces_placement_fen: str):
    pieces = []

    for char in pieces_placement_fen:
        if char.isdigit():
            for i in range(int(char)):
                pieces.append(0b0000)
        elif char == "/":
            continue
        else:
            match char:
                case "P":
                    pieces.append(0b1001)

                case "N":
                    pieces.append(0b1010)

                case "B":
                    pieces.append(0b1011)

                case "R":
                    pieces.append(0b1100)

                case "Q":
                    pieces.append(0b1101)

                case "K":
                    pieces.append(0b1110)

                case "p":
                    pieces.append(0b0001)

                case "n":
                    pieces.append(0b0010)

                case "b":
                    pieces.append(0b0011)

                case "r":
                    pieces.append(0b0100)

                case "q":
                    pieces.append(0b0101)

                case "k":
                    pieces.append(0b0110)

                case _:
                    raise Exception()

    for index, piece in enumerate(pieces):
        if piece != 0:
            position.add_piece_by_int(piece, index)


def set_side_to_move(position: Position, side_to_move_fen: str):
    if side_to_move_fen == "w":
        position.side_to_move = PieceColor.WHITE
    else:
        position.side_to_move = PieceColor.BLACK


def set_castling_rights(position: Position, castling_rights_fen: str):
    if "k" in castling_rights_fen:
        position.castling_rights[CastleEnum.BlackShortCastle] = True
    if "q" in castling_rights_fen:
        position.castling_rights[CastleEnum.BlackLongCastle] = True
    if "K" in castling_rights_fen:
        position.castling_rights[CastleEnum.WhiteShortCastle] = True
    if "Q" in castling_rights_fen:
        position.castling_rights[CastleEnum.WhiteLongCastle] = True


def set_en_passant(position: Position, en_passant_fen: str):
    if en_passant_fen != "-":
        position.en_passant_square = get_num_by_square_name(en_passant_fen)


def get_position_from_fen(fen: str) -> Position:
    fen_splited = fen.split()
    position = Position()

    set_piece_placement(position, fen_splited[0])
    set_side_to_move(position, fen_splited[1])
    set_castling_rights(position, fen_splited[2])
    set_en_passant(position, fen_splited[3])
    position.half_moves = int(fen_splited[4])
    position.current_turn = int(fen_splited[5])

    return position


def get_fen_from_position(position: Position) -> str:
    # Convert the board to FEN piece placement
    fen_rows = []
    empty_count = 0

    for i in range(64):
        piece_code = sum([int(i) for i in position.get_piece_and_color_by_square(i)])

        if piece_code == 0:
            empty_count += 1
        else:
            if empty_count > 0:
                fen_rows.append(str(empty_count))
                empty_count = 0
            fen_rows.append(piece_code_to_fen_char(piece_code))

        if (i + 1) % 8 == 0:
            if empty_count > 0:
                fen_rows.append(str(empty_count))
                empty_count = 0
            if i != 63:
                fen_rows.append("/")

    # Active color
    active_color = 'w' if position.side_to_move == PieceColor.WHITE else 'b'

    # Castling availability
    castling_rights = []
    if position.castling_rights[CastleEnum.WhiteShortCastle]:
        castling_rights.append('K')
    if position.castling_rights[CastleEnum.WhiteLongCastle]:
        castling_rights.append('Q')
    if position.castling_rights[CastleEnum.BlackShortCastle]:
        castling_rights.append('k')
    if position.castling_rights[CastleEnum.BlackLongCastle]:
        castling_rights.append('q')
    castling_rights = ''.join(castling_rights) if castling_rights else '-'

    # En passant target square
    en_passant = '-' if not position.en_passant_square else square_to_name(position.en_passant_square)

    # Halfmove clock and fullmove number
    half_moves = str(position.half_moves)
    full_moves = str(position.current_turn)

    return ' '.join([
        ''.join(fen_rows),
        active_color,
        castling_rights,
        en_passant,
        half_moves,
        full_moves
    ])


def piece_code_to_fen_char(piece_code):
    # Maps internal piece codes to FEN characters
    return {
        0b1001: 'P', 0b1010: 'N', 0b1011: 'B', 0b1100: 'R', 0b1101: 'Q', 0b1110: 'K',
        0b0001: 'p', 0b0010: 'n', 0b0011: 'b', 0b0100: 'r', 0b0101: 'q', 0b0110: 'k'
    }.get(piece_code, None)


def square_to_name(square):
    # Convert square index to algebraic notation (e.g., 0 -> a8)
    file = square % 8
    rank = 8 - (square // 8)
    return f"{chr(file + 97)}{rank}"

File: backend/ChessEngine.Application/FEN/FenGenerator.cs
using ChessEngine.Enums;
using ChessEngine.Helpers;
using ChessEngine.Domain.Models;

namespace ChessEngine.Application.Fen
{
    public static class FenGenerator
    {
        /// <summary>
        /// Sets the piece placement on the board based on the FEN string.
        /// </summary>
        /// <param name="position">The Position object to modify.</param>
        /// <param name="piecesPlacementFen">The piece placement part of the FEN string.</param>
        public static void SetPiecePlacement(Position position, string piecesPlacementFen)
        {
            int square = 0; // Start from square 0 (a8)

            foreach (char c in piecesPlacementFen)
            {
                if (char.IsDigit(c))
                {
                    square += (int)char.GetNumericValue(c); // Skip empty squares
                }
                else if (c == '/')
                {
                    continue; // Move to the next rank
                }
                else
                {
                    int? pieceCode = FenCharToPieceCode(c);
                    if (pieceCode.HasValue)
                    {
                        position.AddPieceByInt(pieceCode.Value, square);
                        square += 1; // Move to the next square
                    }
                    else
                    {
                        throw new ArgumentException($"Invalid FEN character: {c}");
                    }
                }
            }
        }

        /// <summary>
        /// Sets the side to move based on the FEN string.
        /// </summary>
        /// <param name="position">The Position object to modify.</param>
        /// <param name="sideToMoveFen">The side to move part of the FEN string ("w" or "b").</param>
        public static void SetSideToMove(Position position, string sideToMoveFen)
        {
            if (sideToMoveFen == "w")
            {
                position.SideToMove = PieceColor.WHITE;
            }
            else if (sideToMoveFen == "b")
            {
                position.SideToMove = PieceColor.BLACK;
            }
            else
            {
                throw new ArgumentException($"Invalid side to move value in FEN: {sideToMoveFen}");
            }
        }

        /// <summary>
        /// Sets the castling rights based on the FEN string.
        /// </summary>
        /// <param name="position">The Position object to modify.</param>
        /// <param name="castlingRightsFen">The castling rights part of the FEN string.</param>
        public static void SetCastlingRights(Position position, string castlingRightsFen)
        {
            // Reset all castling rights
            foreach (CastleEnum castling in Enum.GetValues(typeof(CastleEnum)))
            {
                position.CastlingRights[castling] = false;
            }

            foreach (char c in castlingRightsFen)
            {
                switch (c)
                {
                    case 'K':
                        position.CastlingRights[CastleEnum.WhiteShortCastle] = true;
                        break;
                    case 'Q':
                        position.CastlingRights[CastleEnum.WhiteLongCastle] = true;
                        break;
                    case 'k':
                        position.CastlingRights[CastleEnum.BlackShortCastle] = true;
                        break;
                    case 'q':
                        position.CastlingRights[CastleEnum.BlackLongCastle] = true;
                        break;
                    default:
                        // Ignore unexpected characters
                        break;
                }
            }
        }

        /// <summary>
        /// Sets the en passant target square based on the FEN string.
        /// </summary>
        /// <param name="position">The Position object to modify.</param>
        /// <param name="enPassantFen">The en passant part of the FEN string.</param>
        public static void SetEnPassant(Position position, string enPassantFen)
        {
            if (enPassantFen != "-")
            {
                position.EnPassantSquare = SquareHelper.GetNumBySquareName(enPassantFen);
            }
            else
            {
                position.EnPassantSquare = null;
            }
        }

        /// <summary>
        /// Parses a FEN string and returns the corresponding Position object.
        /// </summary>
        /// <param name="fen">The full FEN string.</param>
        /// <returns>A Position object representing the chessboard state.</returns>
        public static Position GetPositionFromFen(string fen)
        {
            if (string.IsNullOrWhiteSpace(fen))
                throw new ArgumentException("FEN string cannot be null or empty.", nameof(fen));

            string[] fenParts = fen.Split(' ');
            if (fenParts.Length != 6)
                throw new ArgumentException("Invalid FEN string format.", nameof(fen));

            Position position = new Position();

            // Set piece placement
            SetPiecePlacement(position, fenParts[0]);

            // Set side to move
            SetSideToMove(position, fenParts[1]);

            // Set castling rights
            SetCastlingRights(position, fenParts[2]);

            // Set en passant target square
            SetEnPassant(position, fenParts[3]);

            // Set halfmove clock
            if (!int.TryParse(fenParts[4], out int halfMoves))
                throw new ArgumentException("Invalid halfmove clock in FEN.", nameof(fen));
            position.HalfMoves = halfMoves;

            // Set fullmove number
            if (!int.TryParse(fenParts[5], out int currentTurn))
                throw new ArgumentException("Invalid fullmove number in FEN.", nameof(fen));
            position.CurrentTurn = currentTurn;

            return position;
        }

        /// <summary>
        /// Generates a FEN string from a Position object.
        /// </summary>
        /// <param name="position">The Position object representing the chessboard state.</param>
        /// <returns>A FEN string representing the current position.</returns>
        public static string GetFenFromPosition(Position position)
        {
            if (position == null)
                throw new ArgumentNullException(nameof(position));

            List<string> fenRows = new List<string>();
            int emptyCount = 0;

            for (int i = 0; i < 64; i++)
            {
                var (piece, color) = position.GetPieceAndColorBySquare(i);
                if (piece == PieceEnum.NONE)
                {
                    emptyCount++;
                }  
                else
                {
                    int pieceCode = GetInternalPieceCode(piece, color);
                    if (emptyCount > 0)
                    {
                        fenRows.Add(emptyCount.ToString());
                        emptyCount = 0;
                    }

                    string fenChar = PieceCodeToFenChar(pieceCode);
                    if (fenChar == null)
                        throw new InvalidOperationException($"Invalid piece code at square {i}: {pieceCode}");

                    fenRows.Add(fenChar);
                }

                if ((i + 1) % 8 == 0)
                {
                    if (emptyCount > 0)
                    {
                        fenRows.Add(emptyCount.ToString());
                        emptyCount = 0;
                    }

                    if (i != 63)
                        fenRows.Add("/");
                }
            }

            // Active color
            string activeColor = position.SideToMove == PieceColor.WHITE ? "w" : "b";

            // Castling availability
            List<char> castlingRights = new List<char>();
            foreach (var castling in position.CastlingRights)
            {
                if (castling.Value)
                {
                    switch (castling.Key)
                    {
                        case CastleEnum.WhiteShortCastle:
                            castlingRights.Add('K');
                            break;
                        case CastleEnum.WhiteLongCastle:
                            castlingRights.Add('Q');
                            break;
                        case CastleEnum.BlackShortCastle:
                            castlingRights.Add('k');
                            break;
                        case CastleEnum.BlackLongCastle:
                            castlingRights.Add('q');
                            break;
                    }
                }
            }
            string castlingAvailability = castlingRights.Count > 0 ? new string(castlingRights.ToArray()) : "-";

            // En passant target square
            string enPassant = position.EnPassantSquare.HasValue
                ? SquareHelper.GetSquareNameByNum(position.EnPassantSquare.Value)
                : "-";

            // Halfmove clock and fullmove number
            string halfMoves = position.HalfMoves.ToString();
            string fullMoves = position.CurrentTurn.ToString();

            // Combine all parts
            string fen = string.Join(" ", new string[]
            {
                string.Concat(fenRows),
                activeColor,
                castlingAvailability,
                enPassant,
                halfMoves,
                fullMoves
            });

            return fen;
        }

        /// <summary>
        /// Converts an internal piece code to its corresponding FEN character.
        /// </summary>
        /// <param name="pieceCode">The internal piece code.</param>
        /// <returns>The corresponding FEN character, or null if invalid.</returns>
        private static string PieceCodeToFenChar(int pieceCode)
        {
            var mapping = new Dictionary<int, string>
            {
                {6, "P"}, {7, "N"}, {8, "B"}, {9, "R"}, {10, "Q"}, {11, "K"},
                {0, "p"}, {1, "n"}, {2, "b"}, {3, "r"}, {4, "q"}, {5, "k"}
            };

            return mapping.ContainsKey(pieceCode) ? mapping[pieceCode] : null;
        }

        /// <summary>
        /// Converts a FEN character to its corresponding internal piece code.
        /// </summary>
        /// <param name="c">The FEN character.</param>
        /// <returns>The corresponding internal piece code, or null if invalid.</returns>
        private static int? FenCharToPieceCode(char c)
        {
            var mapping = new Dictionary<char, int>
            {
                {'P', 6}, {'N', 7}, {'B', 8}, {'R', 9}, {'Q', 10}, {'K', 11},
                {'p', 0}, {'n', 1}, {'b', 2}, {'r', 3}, {'q', 4}, {'k', 5}
            };

            return mapping.ContainsKey(c) ? (int?)mapping[c] : null;
        }

        /// <summary>
        /// Converts a piece and its color to the internal piece code.
        /// </summary>
        /// <param name="piece">The type of the piece.</param>
        /// <param name="color">The color of the piece.</param>
        /// <returns>The internal piece code.</returns>
        private static int GetInternalPieceCode(PieceEnum piece, PieceColor color)
        {
            // Assuming the internal piece codes are defined as follows: +++1
            // WHITE_PAWN = 6, WHITE_KNIGHT = 7, WHITE_BISHOP = 8, WHITE_ROOK = 9,
            // WHITE_QUEEN = 10, WHITE_KING = 11
            // BLACK_PAWN = 0, BLACK_KNIGHT = 1, BLACK_BISHOP = 2, BLACK_ROOK = 3,
            // BLACK_QUEEN = 4, BLACK_KING = 5

            if (color == PieceColor.WHITE)
            {
                switch (piece)
                {
                    case PieceEnum.PAWN:
                        return 6;
                    case PieceEnum.KNIGHT:
                        return 7;
                    case PieceEnum.BISHOP:
                        return 8;
                    case PieceEnum.ROOK:
                        return 9;
                    case PieceEnum.QUEEN:
                        return 10;
                    case PieceEnum.KING:
                        return 11;
                    default:
                        throw new ArgumentException($"Invalid piece type: {piece}");
                }
            }
            else // BLACK
            {
                switch (piece)
                {
                    case PieceEnum.PAWN:
                        return 0;
                    case PieceEnum.KNIGHT:
                        return 1;
                    case PieceEnum.BISHOP:
                        return 2;
                    case PieceEnum.ROOK:
                        return 3;
                    case PieceEnum.QUEEN:
                        return 4;
                    case PieceEnum.KING:
                        return 5;
                    default:
                        throw new ArgumentException($"Invalid piece type: {piece}");
                }
            }
        }
    }
}
